### ISSUE TITLE
Migrate other functions into datavalues2

### ISSUE THREAD

**Issue by sundy-li** [datafuselabs/databend#3967]
Now we have a new designed datavalues crate named `datavalues2` in branch `datavalues-dev`.
This new crate has a better design than the previous `datavalue`, including some new features:

1. New datatypes: we do not use enum-based datatype anymore. Instead, we have a new datatype trait `DataType`, and there is `NullableType` which means the datatype can be nullable.

2. New columns: Corresponding to the new data type and we have `ConstColumn` and `NullableColumn`. The `ConstColumn` is a column with constant value, and the `NullableColumn` is a column with an extra bitmap. The other columns are not nullable by default except `NullableColumn`.

3. `ColumnViewer`:  A helper class that can fast access the column without extra cost. What's more, we do not need to care about whether it is nullable or constant. We can use `value(i)` to access the item by index and `null_at(i)` to check it is null or not. The `ColumnViewer` introduces `null_mask` & `non_const_mask` to better handle nullable and constant cases.

4. `ColumnViewerIter`: A helper class to iterator the column. It's a wrapper of `ColumnViewer`

5. `Scalar` types: includes 10 `primitive types`, `boolean`, and `Vec<u8>` types. It's a wonderful idea from  [type-exercise-in-rust](https://github.com/skyzh/type-exercise-in-rust)

6. `ScalarBinaryExpression` && `ScalarUnaryExpression` based on `Scalar` types.

....

Yet, there are a lot of work TBD to make the new datavalues merge into `main` branch.

The main task is functions, I'll mask the migration as `good-first-issue` cause there are many things to do. It may take long time to finish the migration on one's own. And what's more, it will be helpful to understand the new datavalues under migration.

Tasks are:

- [x] #4029

- [x]  datavalues2 && arithmetic functions  [wip @zhyass]

- [x] #3968

- [x] #3969

- [x] #3970

- [x]  datavalues2 && tuple/cast functions (already done)
- [x]  datavalues2 && hash functions (already done)

- [x] #3971

- [x] #3972 [wip @PsiACE ]

- [x] #3973

- [x] #3974

- [x] #3975

- [x] #3976

- [x] #3977

Some tips on migration:

1. In most of the functions, we do not have to care about the nullable cases, you can take the column as not nullable, see [Function2Adapter](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/function2_adapter.rs)  && [ColumnViewer](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/datavalues2/src/columns/viewer.rs) for more details.

2. In most of the functions, we do not have to care about the constant cases, `ColumnViewer` could handle it well. Also, if all the arguments are constant, it will unwrap the constant column and return a constant column. See [Function2Adapter](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/function2_adapter.rs)  for more details.

3. `ScalarBinaryExpression` && `ScalarUnaryExpression` may be two helpful class to cover most function cases.

4. Good examples:  [hashe functions](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/hashes/hash_base.rs)

5. There will be no more implicit casts, we have to explicitly cast the arguments. Errors during the function's eval will be thrown rather than keeping it as null, this is not compatible with the old datavalues.

6. You can take any un-assigned tasks you like, check out from branch `datavalues-dev` to start. don't forget to modify the unit tests and stateless tests. Good Luck!

**Comment by sundy-li** [comment]
Target branch is main after #4074
